Mas i1001 docupdate (#1009)
basho/riak_core: docs/claim-version4.md

Taking this idea one step further, one may imagine a perfect ring with n-val replication, but two of the physical nodes are in the same "location", where location can be a rack or a data center. What if something happens within that location that disconnects two consecutive physical nodes at once? Wouldn't it be nice if one could also take the location into consideration when placing the nodes in the ring, such that the each preflist in the ring conveniently spreads over locations?

We came up with a way to do so.

# Requirements

The solution for placing the nodes in a ring is performed by a so called claim algorithm. The user provides the nodes, the ring size (i.e. the number of vnodes) and the target n-val and from that
a mapping is returned that tries to fulfil the following:

1. the ring has exactly ring size elements.
2. all nodes occur approximately equally often: this is called _balanced_.
   (more precise for a given k, each node appears exactly k or k+1 times)
3. all n-val consecutive vnodes in the ring are in different locations.
4. all n-val consecutive anodes in the ring are on different nodes (which should follow if 3 is true).

> It is expected that once a cluster is inititated the ring size and the target n-val will not normally change, hwoever the number of nodes in the cluster will be changed.

Reviewer: ```suggestion
It is expected that once a cluster is inititated the ring size and the target n-val will not normally change, however the number of nodes in the cluster will be changed.
```

> The first, second and fourth requirements cannot be relaxed. But the third one is not guaranteed to be possible at all. In fact, Bb using a SAT solver we identified

Reviewer: ```suggestion
The first, second and fourth requirements cannot be relaxed. But the third one is not guaranteed to be possible at all. In fact, By using a SAT solver we identified
```

```
n1 n4 n1 n2 n4 n1 n2 n4 n1 n4 n2 n1 n2 n1 n2 n4
```
where `n3` is indeed removed. (This typical example unfortunately requires a lot of transfers).

# Legacy: no locations

The claim algorithms version 1 to 3 that have been used in Riak before, do not consider locations. There the goal is to just consider the n-val for nodes. The new algorithm also
supports that, such that if you have no locations, you can use this newer algorithm. In fact, you can just configure to use this new claim algorithm and run as usual. The module `riak_code_claim_swapping` checks whether you have defined locations and if not, it puts all the nodes in one location.

> Effectively, the `solve` and `update` function are called with `{NVal, 1}` instead of `NVal` as argument, where the second element of the tuple is the location n-val.

Reviewer: ```suggestion
Effectively, the `solve` and `update` function are called with `#{node => NVal, location => 1}` instead of `NVal` as argument.
```

```

> BinRing = riak_core_claim_binring_alg:solve(16, [4], {2,1}).

Reviewer: ```suggestion
BinRing = riak_core_claim_binring_alg:solve(16, [4], #{node => 2, location => 1}).
```
